src/modules/fancyzones/lib/FancyZones.cpp

```diff
@@ -515,49 +524,114 @@ LRESULT CALLBACK FancyZones::s_WndProc(HWND window, UINT message, WPARAM wparam,
+            return TRUE;
+        }
+        DISPLAY_DEVICE displayDevice = { sizeof(displayDevice) };
 
-            bool validMonitor = true;
-            if (EnumDisplayDevices(mi.szDevice, 0, &displayDevice, 1))
+        if (EnumDisplayDevices(mi.szDevice, 0, &displayDevice, 1))
+        {
+            if (WI_IsFlagSet(displayDevice.StateFlags, DISPLAY_DEVICE_MIRRORING_DRIVER))
             {
-                if (WI_IsFlagSet(displayDevice.StateFlags, DISPLAY_DEVICE_MIRRORING_DRIVER))
-                {
-                    validMonitor = FALSE;
-                }
-                else if (displayDevice.DeviceID[0] != L'\0')
-                {
-                    deviceId = displayDevice.DeviceID;
-                }
+                return TRUE;
             }
+        }
+        if (displayDevice.DeviceID[0] == L'\0')
+        {
+            return TRUE;
+        }
 
-            if (validMonitor)
-            {
-                if (!deviceId)
-                {
-                    deviceId = GetSystemMetrics(SM_REMOTESESSION) ?
-                        L"\\\\?\\DISPLAY#REMOTEDISPLAY#" :
-                        L"\\\\?\\DISPLAY#LOCALDISPLAY#";
-                }
+        const auto capturedScope = reinterpret_cast<localScopeCapture *>(captured_ptr);
+        auto & newDevice = capturedScope->devices.emplace_back(device{L"FallbackDevice", monitor});
 
-                auto strongThis = reinterpret_cast<FancyZones*>(data);
-                strongThis->AddZoneWindow(monitor, deviceId);
-            }
+        std::array<wchar_t, 256> parsedId{};
+        if (!ParseDeviceId(displayDevice.DeviceID, parsedId.data(), size(parsedId)))
+        {
+            return TRUE;
+        }
+        std::wstring_view parsedIdView{parsedId.data(), wcslen(parsedId.data())};
+        newDevice._deviceID = parsedIdView;
+        if (parsedIdView.find(L'#') != std::wstring::npos)
+        {
+            // Check if EnumDisplayMonitors returned a duplicate data, it's an API bug and we should workaround it
+            const auto [_, hasUniqueID] = capturedScope->discoveredDeviceIDs.emplace(parsedIdView);
+            capturedScope->duplicateDeviceIdFound = !hasUniqueID || capturedScope->duplicateDeviceIdFound;
         }
         return TRUE;
     };
+    EnumDisplayMonitors(nullptr, nullptr, enumDisplayMonitorsCallback, reinterpret_cast<LPARAM>(&localScope));
 
+    if (GetSystemMetrics(SM_REMOTESESSION))
     {
-        std::unique_lock writeLock(m_lock);
-        m_zoneWindowMap.clear();
+        // When running in a remote session, devices have "Default_Monitor#<transient_part>" DeviceIDS.
+        // Since a remote session can have multiple devices, we can't just assign them a single "REMOTE" ID.
+        // We'll assign them unique numeric IDs instead.
+        for (size_t i = 0; i < size(localScope.devices); ++i)
+        {
+            wchar_t remoteDeviceID[32];
+            swprintf_s(remoteDeviceID, L"REMOTE%zu", i);

```

Comment:
Let's add an underscore after `REMOTE`  -> `L"REMOTE_%zu"`, since we want to make all the settings data as readable as possible.